sdk/go/pulumi/alias.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Alias is a partial description of prior named used for a resource. It can be processed in the
+// context of a resource creation to determine what the full aliased URN would be.
+type Alias struct {
+	// The URN that uniquely identifies a resource
+	URN URNInput
+	// The previous name of the resource.  If not provided, the current name of the resource is used.
+	Name StringInput
+	// The previous type of the resource.  If not provided, the current type of the resource is used.
+	Type StringInput
+	// The previous parent of the resource.  If not provided, the current parent of the resource is used.
+	Parent Resource
+	// The previous parent of the resource in URN format, mutually exclusive to 'Parent'
+	ParentURN URNInput
+	// The name of the previous stack of the resource.  If not provided, defaults to `context.GetStack()
+	Stack StringInput
+	// The previous project of the resource. If not provided, defaults to `context.GetProject()`.
+	Project StringInput
+}
+
+func (a *Alias) isCollapsed() bool {
+	return a.URN != nil
+}
```

Comment:
Why is this notion of a modal `collapsed` state needed?  

In the TypeScript SDK we just store the raw `string` in the parent aliases list.  The idea is that `Alias` is a user convenience input option - but the actual state we manage is purely in terms of URNs.  Is there a reason we can't/shouldn't do that here?  The current approach seems prone to using `Aliases` in the "wrong state".

Reply:
This has to do with a lack of union support in go. We use this type to represent that you can either pass in a URN (string), or an Alias object. We collapse the information provided into a URN that gets written into alias.URN. 

If you have an idea for an alternative approach that would be safer, I'm all ears.

Reply:
Accepting a union makes sense. But internally we should store just the `URN`. And if we do that we don’t need this statefulness in the input struct. 

Also - I believe there is a standard pattern for supporting unions in our input model - we use it for `oneOf` in our schema codegen.

Reply:
My understanding is that we use this to represent union types in other languages, but don't actually use this type in our go SDK implementation. 

@pgavlin can you provide some input here?

Reply:
> internally we should store just the URN

How does TS implement inherited aliases using just the URN?

Reply:
@pgavlin not exactly sure I follow your question - but the answer is presumably at:

https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/resource.ts#L265

And

https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/resource.ts#L309

It feels like the Go implementation can/should be very similar to that - especially in the state that is managed.

Reply:
This is really what I was looking for: https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/resource.ts#L267-L269

Reply:
I agree with @lukehoban on this point. We can just store the collapsed URNs, and `inheritedChildAlias` can be written in terms of the parent URN rather than the parent `Alias`.

Reply:
@lukehoban @pgavlin Help me understand if there's a better path forward here:

Currently, we initialize `resourceState` in the sync path of read/registerResource. This is where we store the aliases:
https://github.com/pulumi/pulumi/blob/2d03df5c97377847c5c020a9b279b1eb9f7517a8/sdk/go/pulumi/context.go#L280-L285

All awaiting/resolution (the part where we get the fully qualified string URNs) happens in the async path in the go func. 

https://github.com/pulumi/pulumi/blob/2d03df5c97377847c5c020a9b279b1eb9f7517a8/sdk/go/pulumi/context.go#L286-L304

The python and node code paths basically do the same thing. The only difference is that they overwrite the `resourceState `equivalent with the resolved `string` values at the end. As an alternative, I could replace all the aliases with `&Alias{ URN: resolved }` so that none of the other properties are around to cause confusion. I could also create another private property on `resourceState` like `resolvedAliases` that I populate in the async path.

I'm not really sure what problem we're trying to avoid here. This state is completely internal, and simply an implementation detail of the feature. Why are we concerned about it?

Reply:
@EvanBoyle here's what I'd recommend doing: https://github.com/pulumi/pulumi/compare/evan/goAliases..pgavlin/goAliases

These changes avoid the statefulness in `Alias` and let the internal bits of the SDK work entirely in terms of URNs.

Reply:
Ah, thanks for that! It looks much cleaner. I had to make a small update, but I've pushed the changes and the tests are all passing.